src/OpenTelemetry.Exporter.Prometheus/PrometheusExporter.cs

```diff
@@ -36,18 +36,53 @@ public class PrometheusExporter : MetricExporter
         public PrometheusExporter(PrometheusExporterOptions options)
         {
             this.Options = options;
-            this.Metrics = new List<Metric>();
+            this.LongMetrics = new List<Metric<long>>();
+            this.DoubleMetrics = new List<Metric<double>>();
         }
 
-        internal List<Metric> Metrics { get; private set; }
+        private List<Metric<long>> LongMetrics { get; set; }
+
+        private List<Metric<double>> DoubleMetrics { get; set; }
 
         /// <inheritdoc/>
-        public override Task<ExportResult> ExportAsync(List<Metric> metrics, CancellationToken cancellationToken)
+        public override Task<ExportResult> ExportAsync<T>(List<Metric<T>> metrics, CancellationToken cancellationToken)
         {
-            // Prometheus uses a pull process, not a push
-            // Store the updated metrics internally for the next pull and return success.
-            this.Metrics = metrics;
+            // Prometheus uses a pull model, not a push.
+            // Accumulate the exported metrics internally, return success.
+            // The pull process will read this internally stored metrics
+            // at its own schedule.
+            if (typeof(T) == typeof(double))
+            {
+                var doubleList = metrics
+                .Select(x => (x as Metric<double>))
+                .ToList();
+
+                this.DoubleMetrics.AddRange(doubleList);
+            }
+            else
+            {
+                var longList = metrics
+                .Select(x => (x as Metric<long>))
+                .ToList();
+
+                this.LongMetrics.AddRange(longList);
+            }
+
             return Task.FromResult(ExportResult.Success);
         }
+
+        internal List<Metric<long>> GetAndClearLongMetrics()
```

Comment:
It could be dangerous if we clear the list before we've successfully written the data. eg exception thrown before written to http stream.

Not sure if this needs to be addressed now, or tracked in another issue to make the exporter more resilient.

Reply:
Agree.  This need to be addressed. I will defer prometheus related improvements to separate PR as Hananiel has agreed to pick that work after this is merged.